File: apps/web/src/lib/settings/onboarding.test.ts
import { test } from 'node:test';
import assert from 'node:assert/strict';

import {
	ONBOARDING_STEPS,
	ONBOARDING_TOTAL_STEPS,
	visibleOnboardingSteps,
	PRIMARY_GOAL_KEY,
	PRIMARY_GOAL_VALUES,
	planPresetForGoal,
} from './onboarding';
import { en } from '../i18n/locales/en';

test('PRIMARY_GOAL_KEY is the universal-prefs bag key the wizard writes', () => {
	// Hard-coded by name across the wizard, the Settings nudge, and
	// (eventually) the post-onboarding plan-suggestion surface. A
	// rename here without grepping every consumer would silently
	// orphan the saved value.
	assert.equal(PRIMARY_GOAL_KEY, 'primary_goal');
});

test('every PRIMARY_GOAL_VALUE has a non-empty onboarding.goal i18n label', () => {
	// Labels are localized (key `onboarding.goal.<value>`), not held as
	// English strings in the helper. The wizard renders `m(...)` for
	// each value, so a value without a matching key would render the raw
	// key fallback. Pin the en catalogue (the canonical superset that
	// messages_parity.test.ts forces every other locale to mirror).
	const labels = en as Record<string, string>;
	for (const v of PRIMARY_GOAL_VALUES) {
		const key = `onboarding.goal.${v}`;
		assert.ok(
			labels[key] && labels[key].trim().length > 0,
			`PRIMARY_GOAL_VALUES contains "${v}" but en has no "${key}" label`,
		);
	}
});

test('PRIMARY_GOAL_VALUES includes the four GoalEvent distances + 2 non-distance goals', () => {
	// The four distance goals map 1:1 to `training.ts#GoalEvent` so
	// a future plan-creation flow can translate without a mapping
	// table. Pin the membership explicitly.
	assert.ok((PRIMARY_GOAL_VALUES as readonly string[]).includes('5k'));
	assert.ok((PRIMARY_GOAL_VALUES as readonly string[]).includes('10k'));
	assert.ok((PRIMARY_GOAL_VALUES as readonly string[]).includes('half_marathon'));
	assert.ok((PRIMARY_GOAL_VALUES as readonly string[]).includes('marathon'));
	assert.ok((PRIMARY_GOAL_VALUES as readonly string[]).includes('general_fitness'));
	assert.ok((PRIMARY_GOAL_VALUES as readonly string[]).includes('weight_loss'));
	// And exactly those 6 — guards against a silent addition.
	assert.equal(PRIMARY_GOAL_VALUES.length, 6);
});

test('planPresetForGoal maps distance goals 1:1 and seeds beginners into walk-run 5K', () => {
	// Distance goals map straight to GoalEvent, no beginner toggle.
	assert.deepEqual(planPresetForGoal('10k'), { goalEvent: 'distance_10k', beginnerWalkRun: false });
	assert.deepEqual(planPresetForGoal('half_marathon'), { goalEvent: 'distance_half', beginnerWalkRun: false });
	assert.deepEqual(planPresetForGoal('marathon'), { goalEvent: 'distance_full', beginnerWalkRun: false });
	// The three beginner-leaning goals all land on the walk-run 5K — the
	// affordance a brand-new runner would otherwise never find.
	assert.deepEqual(planPresetForGoal('5k'), { goalEvent: 'distance_5k', beginnerWalkRun: true });
	assert.deepEqual(planPresetForGoal('general_fitness'), { goalEvent: 'distance_5k', beginnerWalkRun: true });
	assert.deepEqual(planPresetForGoal('weight_loss'), { goalEvent: 'distance_5k', beginnerWalkRun: true });
});

test('planPresetForGoal covers every PRIMARY_GOAL_VALUE', () => {
	// No goal falls through to an undefined preset.
	for (const v of PRIMARY_GOAL_VALUES) {
		const preset = planPresetForGoal(v);
		assert.ok(preset && typeof preset.goalEvent === 'string');
	}
});

test('ONBOARDING_TOTAL_STEPS matches the wizard step count', () => {
	// Drives the progress-dot indicator. Drift would mean the user
	// sees "step 4 of 7" while there are 8 actual steps, or vice
	// versa. Pin the constant; the page-level test counts the
	// rendered <section> blocks.
	assert.equal(ONBOARDING_TOTAL_STEPS, 7);
});

test('the full wizard is the seven steps, in order', () => {
	assert.deepEqual(
		[...ONBOARDING_STEPS],
		['name', 'units', 'goal', 'about', 'privacy', 'notifications', 'done'],
	);
	assert.equal(ONBOARDING_TOTAL_STEPS, ONBOARDING_STEPS.length);
});

test('the notifications step appears only when push can be turned on here', () => {
	const withPush = visibleOnboardingSteps({ supported: true, permission: 'default', subscribed: false });
	assert.deepEqual(withPush, [...ONBOARDING_STEPS]);

	for (const [why, push] of [
		['no browser support or no deployed key', { supported: false, permission: 'default', subscribed: false }],
		['blocked in the browser', { supported: true, permission: 'denied', subscribed: false }],
		['no Notification API at all', { supported: true, permission: 'unsupported', subscribed: false }],
		['already on for this device', { supported: true, permission: 'granted', subscribed: true }],
	] as const) {
		const steps = visibleOnboardingSteps(push);
		assert.ok(!steps.includes('notifications'), why);
		assert.equal(steps.length, ONBOARDING_TOTAL_STEPS - 1, why);
		assert.equal(steps[steps.length - 1], 'done', `${why}: the wizard still ends on done`);
	}
});

test('granted but not yet subscribed still offers the step', () => {
	// Permission granted to the site does not mean this device is subscribed.
	const steps = visibleOnboardingSteps({ supported: true, permission: 'granted', subscribed: false });
	assert.ok(steps.includes('notifications'));
});

File: apps/mobile_android/test/onboarding_test.dart
import 'package:api_client/api_client.dart';
import 'package:flutter_test/flutter_test.dart';

import '../lib/onboarding.dart';
import '../lib/training.dart';

/// Pure-contract twin of web `apps/web/src/lib/settings/onboarding.test.ts`.
/// Keeps the mobile setup-wizard enum + step count in lockstep with web.
void main() {
  group('onboarding contract', () {
    test('primaryGoalKey is the universal-prefs bag key the wizard writes', () {
      expect(primaryGoalKey, 'primary_goal');
      // And it's the SAME constant the SettingsKeys registry exposes, so a
      // rename can't silently split the writer from the reader.
      expect(primaryGoalKey, SettingsKeys.primaryGoal);
    });

    test('primaryGoalValues covers the fixed six-value enum, in order', () {
      expect(primaryGoalValues, [
        'general_fitness',
        'weight_loss',
        '5k',
        '10k',
        'half_marathon',
        'marathon',
      ]);
    });

    test('primaryGoalValues includes the four distance goals + two non-distance',
        () {
      for (final v in ['5k', '10k', 'half_marathon', 'marathon']) {
        expect(primaryGoalValues.contains(v), isTrue, reason: 'missing $v');
      }
      expect(primaryGoalValues.contains('general_fitness'), isTrue);
      expect(primaryGoalValues.contains('weight_loss'), isTrue);
      expect(primaryGoalValues.length, 6);
    });

    test('onboardingTotalSteps matches web (7)', () {
      expect(onboardingTotalSteps, 7);
      // The constant is written out (a list length is not a constant
      // expression), so pin it against the list it describes.
      expect(setupWizardSteps.length, onboardingTotalSteps);
    });

    test('setupWizardSteps are web\'s ONBOARDING_STEPS ids, in order', () {
      expect(setupWizardSteps, [
        'name',
        'units',
        'goal',
        'about',
        'run-privacy',
        'notifications',
        'done',
      ]);
    });

    test('visibleSetupWizardSteps drops run-privacy once the launch flow '
        'already asked, and nothing else', () {
      final asked = visibleSetupWizardSteps(privacyAlreadyChosen: true);
      expect(asked.contains('run-privacy'), isFalse);
      expect(asked.length, onboardingTotalSteps - 1);
      expect(asked, setupWizardSteps.where((s) => s != 'run-privacy'));
      // Order is preserved — the wizard indexes this list.
      expect(asked.first, 'name');
      expect(asked.last, 'done');
    });

    test('visibleSetupWizardSteps walks every step when privacy is unanswered',
        () {
      expect(visibleSetupWizardSteps(privacyAlreadyChosen: false),
          setupWizardSteps);
    });

    test('planPresetForGoal maps distance goals 1:1 and seeds beginners into '
        'walk-run 5K', () {
      expect(planPresetForGoal('10k').goalEvent, GoalEvent.distance10k);
      expect(planPresetForGoal('10k').beginnerWalkRun, isFalse);
      expect(planPresetForGoal('half_marathon').goalEvent, GoalEvent.distanceHalf);
      expect(planPresetForGoal('marathon').goalEvent, GoalEvent.distanceFull);
      for (final v in ['5k', 'general_fitness', 'weight_loss']) {
        expect(planPresetForGoal(v).goalEvent, GoalEvent.distance5k,
            reason: '$v should seed the 5K');
        expect(planPresetForGoal(v).beginnerWalkRun, isTrue,
            reason: '$v should tick the walk-run toggle');
      }
    });

    test('ApiClient.dateOnly emits a zone-free YYYY-MM-DD calendar day', () {
      // A DOB is a calendar day, not an instant — the wizard must not let
      // a user east/west of UTC roll a day. Pad single digits.
      expect(ApiClient.dateOnly(DateTime(2001, 2, 3)), '2001-02-03');
      expect(ApiClient.dateOnly(DateTime(1990, 11, 25)), '1990-11-25');
      // A late-evening local instant keeps its local calendar day (no UTC
      // conversion that would advance to the next date).
      expect(ApiClient.dateOnly(DateTime(2001, 2, 3, 23, 30)), '2001-02-03');
    });
  });
}
